crates/libs/windows/src/core/heap.rs

```diff
@@ -32,20 +32,42 @@ pub unsafe fn heap_free(ptr: RawPtr) {
     }
 }
 
-/// Copy a slice of `T` into a freshly allocated buffer with an additional default `T` at the end.
+/// Copy an iterator of `T` into a freshly allocated buffer with an additional default `T` at the end.
 ///
-/// Returns a pointer to the beginning of the buffer
+/// Returns a pointer to the beginning of the buffer. This pointer must be freed when done using `heap_free`.
 ///
 /// # Panics
 ///
-/// This function panics if the heap allocation fails or if the pointer returned from
-/// the heap allocation is not properly aligned to `T`.
-pub fn heap_string<T: Copy + Default + Sized>(slice: &[T]) -> *const T {
-    unsafe {
-        let buffer = heap_alloc((slice.len() + 1) * std::mem::size_of::<T>()).expect("could not allocate string") as *mut T;
-        assert!(buffer.align_offset(std::mem::align_of::<T>()) == 0, "heap allocated buffer is not properly aligned");
-        buffer.copy_from_nonoverlapping(slice.as_ptr(), slice.len());
-        buffer.add(slice.len()).write(T::default());
-        buffer
+/// This function panics if the heap allocation fails, the alignment requirements of 'T' surpass
+/// 8 (HeapAlloc's alignment) or if len is less than the number of items in the iterator.
+pub fn string_from_iter<I, T>(iter: I, len: usize) -> *const T
+where
+    I: Iterator<Item = T>,
+    T: Copy + Default,
+{
+    // alignment of memory returned by HeapAlloc is at least 8
+    // Source: https://docs.microsoft.com/en-us/windows/win32/api/heapapi/nf-heapapi-heapalloc
+    // Ensure that T has sufficient alignment requirements
+    assert!(std::mem::align_of::<T>() <= 8, "T alignment surpasses HeapAlloc alignment");
+
+    let len = len + 1;
+    let ptr = heap_alloc(len * std::mem::size_of::<T>()).expect("could not allocate string") as *mut T;
+    let mut encoder = iter.chain(core::iter::once(T::default()));
+
+    for i in 0..len {
+        // SAFETY: ptr points to an allocation object of size `len`, indices accessed are always lower than `len`
+        unsafe {
+            core::ptr::write(
+                ptr.add(i),
+                match encoder.next() {
+                    Some(encoded) => encoded,
+                    None => break,
```

Comment:
It might be useful to do the following here instead of the assert in code a few lines later:

```rust
debug_assert!(i == len -1);
```

Essentially, while this code is always safe (i.e., we'll never try to write to unallocated memory), if the iterator's length and `len` don't match we either end up allocating too little memory or too much. It seems reasonable to help user's of this function not make that mistake.

Reply:
Do you mean `debug_assert!(i < len)` and not `==`?

Reply:
I meant `==`, because I believe we expect that length will always be equal to the number of elements in the iterator.